## - Creating <font color='blue'>m18_Global_QLQ30groups65gl</font> feature based on M18_Global_QLQ30
## - Using <u>most discriminating</u> features from <font color='blue'>m0</font> and <font color='blue'>m3</font>
## - SVM - Support Vector Machine
## - SMOTE and without SMOTE 

In [55]:
#importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import pyreadstat

In [56]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [57]:
#reading data from SPSS file
df=pd.read_spss(r"C:\Users\Chris\Desktop\Cancer_Prediction\BOUNCE_Full Dataset_m0_m3_m6_m9_m12_m15_m18_09-12-2021_ORIGINAL.sav")

In [58]:
# create a list of our conditions (breakpoint 65)
# >65  ->  0
# <65  ->  1
conditions = [(df['M18_Global_QLQ30'] >= 0) & (df['M18_Global_QLQ30'] <= 65),(df['M18_Global_QLQ30'] > 65) & (df['M18_Global_QLQ30'] <= 100)]

In [59]:
# create a list of the values we want to assign for each condition
values = [9, 8]

In [60]:
df['m18_Global_QLQ30groups65gl'] = np.select(conditions, values)

C:\Users\Chris\AppData\Local\Temp/ipykernel_10016/2152058818.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df['m18_Global_QLQ30groups65gl'] = np.select(conditions, values)


In [61]:
df

mrn  FORTH_id  M12_data      MissingM0  \
0                 1/2019 MJF       1.0       1.0  complete data   
1                10/2019 ICM       2.0       1.0  complete data   
2               100/2019 AIM       3.0       0.0  complete data   
3               101/2019 MDG       4.0       1.0  complete data   
4               103/2019 PCP       5.0       1.0  complete data   
..                       ...       ...       ...            ...   
761                  EIEO117     832.0       0.0            NaN   
762                  EIEO197     833.0       1.0            NaN   
763                  EIEO225     834.0       0.0            NaN   
764             154/2019 ACL     836.0       NaN            NaN   
765  Rabin Medical Center_50     837.0       NaN            NaN   

                   M0_MLmodels      misclasssifiedM0  misclasssifiedM0_012  \
0    used in modelling (N=609)  correctly classified  correctly classified   
1    used in modelling (N=609)  correctly classified  correctly classified   
2    used in modelling (N=609)  correctly classified  correctly classified   
3    used in modelling (N=609)  correctly classified  correctly classified   
4    used in modelling (N=609)  correctly classified  correctly classified   
..                         ...                   ...                   ...   
761                        NaN                   NaN                   NaN   
762                        NaN                   NaN                   NaN   
763                        NaN                   NaN                   NaN   
764                        NaN                   NaN                   NaN   
765                        NaN                   NaN                   NaN   

                       age.date_filter   age care_team  ...  m18_anxiety_HADS  \
0    40-70 years old within date range  66.0     CHAMP  ...          0.428571   
1    40-70 years old within date range  54.0     CHAMP  ...          0.000000   
2    40-70 years old within date range  45.0     CHAMP  ...               NaN   
3    40-70 years old within date range  70.0     CHAMP  ...          0.000000   
4    40-70 years old within date range  49.0     CHAMP  ...          1.285714   
..                                 ...   ...       ...  ...               ...   
761          outside age range or date   NaN       IEO  ...          1.000000   
762  40-70 years old within date range  43.0       IEO  ...               NaN   
763          outside age range or date   NaN       IEO  ...               NaN   
764                                NaN   NaN     CHAMP  ...          0.000000   
765                                NaN   NaN      HUJI  ...          0.142857   

    m18_depression_HADS  m18_mental_health_total_HADS  \
0              0.000000                      0.214286   
1              0.428571                      0.214286   
2                   NaN                           NaN   
3              0.000000                      0.000000   
4              1.142857                      1.214286   
..                  ...                           ...   
761            0.285714                      0.642857   
762                 NaN                           NaN   
763                 NaN                           NaN   
764            0.142857                      0.071429   
765            0.000000                      0.071429   

    m18_negative_affect_PANAS  m18_positive_affect_PANAS m18_cancer_stage  \
0                         1.2                        4.0           Stage1   
1                         1.2                        4.4          Stage 2   
2                         NaN                        NaN          Stage 3   
3                         1.0                        3.8              NaN   
4                         2.0                        3.0              NaN   
..                        ...                        ...              ...   
761                       3.2                        4.0           Stage1   
762      

In [62]:
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(0, np.NaN)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(9, 1)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(8, 0)

In [63]:
df

mrn  FORTH_id  M12_data      MissingM0  \
0                 1/2019 MJF       1.0       1.0  complete data   
1                10/2019 ICM       2.0       1.0  complete data   
2               100/2019 AIM       3.0       0.0  complete data   
3               101/2019 MDG       4.0       1.0  complete data   
4               103/2019 PCP       5.0       1.0  complete data   
..                       ...       ...       ...            ...   
761                  EIEO117     832.0       0.0            NaN   
762                  EIEO197     833.0       1.0            NaN   
763                  EIEO225     834.0       0.0            NaN   
764             154/2019 ACL     836.0       NaN            NaN   
765  Rabin Medical Center_50     837.0       NaN            NaN   

                   M0_MLmodels      misclasssifiedM0  misclasssifiedM0_012  \
0    used in modelling (N=609)  correctly classified  correctly classified   
1    used in modelling (N=609)  correctly classified  correctly classified   
2    used in modelling (N=609)  correctly classified  correctly classified   
3    used in modelling (N=609)  correctly classified  correctly classified   
4    used in modelling (N=609)  correctly classified  correctly classified   
..                         ...                   ...                   ...   
761                        NaN                   NaN                   NaN   
762                        NaN                   NaN                   NaN   
763                        NaN                   NaN                   NaN   
764                        NaN                   NaN                   NaN   
765                        NaN                   NaN                   NaN   

                       age.date_filter   age care_team  ...  m18_anxiety_HADS  \
0    40-70 years old within date range  66.0     CHAMP  ...          0.428571   
1    40-70 years old within date range  54.0     CHAMP  ...          0.000000   
2    40-70 years old within date range  45.0     CHAMP  ...               NaN   
3    40-70 years old within date range  70.0     CHAMP  ...          0.000000   
4    40-70 years old within date range  49.0     CHAMP  ...          1.285714   
..                                 ...   ...       ...  ...               ...   
761          outside age range or date   NaN       IEO  ...          1.000000   
762  40-70 years old within date range  43.0       IEO  ...               NaN   
763          outside age range or date   NaN       IEO  ...               NaN   
764                                NaN   NaN     CHAMP  ...          0.000000   
765                                NaN   NaN      HUJI  ...          0.142857   

    m18_depression_HADS  m18_mental_health_total_HADS  \
0              0.000000                      0.214286   
1              0.428571                      0.214286   
2                   NaN                           NaN   
3              0.000000                      0.000000   
4              1.142857                      1.214286   
..                  ...                           ...   
761            0.285714                      0.642857   
762                 NaN                           NaN   
763                 NaN                           NaN   
764            0.142857                      0.071429   
765            0.000000                      0.071429   

    m18_negative_affect_PANAS  m18_positive_affect_PANAS m18_cancer_stage  \
0                         1.2                        4.0           Stage1   
1                         1.2                        4.4          Stage 2   
2                         NaN                        NaN          Stage 3   
3                         1.0                        3.8              NaN   
4                         2.0                        3.0              NaN   
..                        ...                        ...              ...   
761                       3.2                        4.0           Stage1   
762      

## Dataset which includes only <font color='blue'>m0</font> and <font color='blue'>m3</font> feautres 

In [64]:
df = df[['M0_employment_status','M0_drinking','M0_diet','M0_MedDiet','M0_SpecialDiet','M0_weight','M0_bmi',
         'M0_obese','M0_exercise','M0_life_events_1.2','M0_sick_leave_days','M0_extraversion_TIPI',
         'M0_aggreableness_TIPI','M0_conscientiousness_TIPI','M0_emot_stab_TIPI','M0_openess_TIPI',
         'M0_optimism_LOT','M0_comprehensibility_SOC','M0_manageability_SOC','M0_meaningfulness_SOC',
         'M0_Forward_PACT','M0_Trauma_PACT','M0_Flexibility_PACT','M0_self_blame_CERQ','M0_other_blame_CERQ',
         'M0_rumination_CERQ','M0_catastrophizing_CERQ','M0_perspective_CERQ','M0_pos_refus_CERQ',
         'M0_pos_reapp_CERQ','M0_acceptance_CERQ','M0_planning_CERQ','M0_negative_overall_CERQ',
         'M0_positive_overall_CERQ','M0_mindfulness_MAAS','M0_resilience_CDRISC','M0_coping_with_cancer_CBI',
         'M0_fear_of_recur_FCRI','M0_anxiety_HADS','M0_depression_HADS','M0_mental_health_total_HADS',
         'M0_positive_affect_PANAS','M0_negative_affect_PANAS','M0_Global_QLQ30','M0_Phys_Fun_QLQ30',
         'M0_Role_Fun_QLQ30','M0_Emot_Fun_QLQ30','M0_Cogn_Fun_QLQ30','M0_Soc_Fun_QLQ30','M0_Fatigue_QLQ30',
         'M0_Nausea_QLQ30','M0_Pain_QLQ30','M0_Dyspnoea_QLQ30','M0_Insomnia_QLQ30','M0_Apetite_QLQ30',
         'M0_Constipation_QLQ30','M0_Diarrhoea_QLQ30','M0_Financial_QLQ30','M0_Body_Image_BR23',
         'M0_Side_Effects_BR23','M0_Breast_Symptoms_BR23','M0_Arm_Symptoms_BR23','M0_Upset_Hair_Image_BR23',
         'M0_Future_Persp_Image_BR23','M0_Sex_Funct_BR23','M0_Sex_Enjoy_BR23','M0_performancestatus',
         'm0_menopausalstatuspre','M0_surgery','M0_chemo','M0_radio','M0_psychotropics',
         'm0_Global_QLQ30groups65gl','m0_Phys_Fun_QLQ30groups80gl','diagnosis_M0_delay','m3_life_events_1.2',
         'm3_sick_leave_days','m3_employer_support','m3_domestic_help',
         'm3_perceived_suppport_1_item','m3_nccn_distress_thermometer',
         'm3_general_se_1_item','m3_rate_how_much_your_faith_including_religiosity_and_spiri',
         'M3_instumental_support','M3_emotional_support','M3_social_support_total','M3_PTGI_relating_to_others',
         'M3_PTGI_new_possibilities','M3_PTGI_personal_strength','M3_PTGI_spiritual_change','M3_PTGI_appreciation_of_life',
         'M3_PTGI_total_score','M3_negative_affect_PANAS','M3_positive_affect_PANAS','M3_anxiety_HADS',
         'M3_depression_HADS','M3_mental_health_total_HADS','M3_Global_QLQ30','M3_Phys_Fun_QLQ30',
         'M3_Role_Fun_QLQ30','M3_Emot_Fun_QLQ30','M3_Cogn_Fun_QLQ30','M3_Soc_Fun_QLQ30','M3_Fatigue_QLQ30',
         'M3_Nausea_QLQ30','M3_Pain_QLQ30','M3_Dyspnoea_QLQ30','M3_Insomnia_QLQ30','M3_Apetite_QLQ30',
         'M3_Constipation_QLQ30','M3_Diarrhoea_QLQ30','M3_Financial_QLQ30','M3_Body_Image_m3_br23',
         'M3_Side_Effects_m3_br23','M3_Breast_Symptoms_m3_br23','M3_Arm_Symptoms_m3_br23','M3_Upset_Hair_Image_m3_br23',
         'M3_Future_Persp_Image_m3_br23','M3_Sex_Funct_m3_br23','M3_Sex_Enjoy_m3_br23','m3_performancestatus',
         'M3_surgery','M3_chemo','M3_radio','M3_psychotropics','M3_Mental_health_support',
         'm3_illness_events0.1.2','m3_Global_QLQ30groups65gl','m3_Phys_Fun_QLQ30groups80gl','m18_Global_QLQ30groups65gl']]

In [65]:
#check the category features
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 130 columns):
 #    Column                                                       Dtype   
---   ------                                                       -----   
 0    M0_employment_status                                         category
 1    M0_drinking                                                  category
 2    M0_diet                                                      category
 3    M0_MedDiet                                                   category
 4    M0_SpecialDiet                                               category
 5    M0_weight                                                    float64 
 6    M0_bmi                                                       float64 
 7    M0_obese                                                     category
 8    M0_exercise                                                  category
 9    M0_life_events_1.2                                  

In [66]:
#convert category to numeric value
from sklearn.preprocessing import LabelEncoder
labelencoder_Y = LabelEncoder()
df.iloc[:,0] = labelencoder_Y.fit_transform(df.iloc[:,0].values)
df.iloc[:,1] = labelencoder_Y.fit_transform(df.iloc[:,1].values)
df.iloc[:,2] = labelencoder_Y.fit_transform(df.iloc[:,2].values)
df.iloc[:,3] = labelencoder_Y.fit_transform(df.iloc[:,3].values)
df.iloc[:,4] = labelencoder_Y.fit_transform(df.iloc[:,4].values)
df.iloc[:,7] = labelencoder_Y.fit_transform(df.iloc[:,7].values)
df.iloc[:,8] = labelencoder_Y.fit_transform(df.iloc[:,8].values)
df.iloc[:,9] = labelencoder_Y.fit_transform(df.iloc[:,9].values)
df.iloc[:,67] = labelencoder_Y.fit_transform(df.iloc[:,67].values)
df.iloc[:,71] = labelencoder_Y.fit_transform(df.iloc[:,71].values)
df.iloc[:,72] = labelencoder_Y.fit_transform(df.iloc[:,72].values)
df.iloc[:,73] = labelencoder_Y.fit_transform(df.iloc[:,73].values)
df.iloc[:,75] = labelencoder_Y.fit_transform(df.iloc[:,75].values)
df.iloc[:,77] = labelencoder_Y.fit_transform(df.iloc[:,77].values)
df.iloc[:,78] = labelencoder_Y.fit_transform(df.iloc[:,78].values)
df.iloc[:,124] = labelencoder_Y.fit_transform(df.iloc[:,124].values)
df.iloc[:,125] = labelencoder_Y.fit_transform(df.iloc[:,125].values)
df.iloc[:,126] = labelencoder_Y.fit_transform(df.iloc[:,126].values)
df.iloc[:,127] = labelencoder_Y.fit_transform(df.iloc[:,127].values)
df.iloc[:,128] = labelencoder_Y.fit_transform(df.iloc[:,128].values)

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [67]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                       0            1        1           1               0   
1                       0            1        1           1               0   
2                       4            1        1           1               0   
3                       2            2        1           1               0   
4                       2            1        0           0               0   
..                    ...          ...      ...         ...             ...   
761                     6            3        3           2               2   
762                     4            1        0           0               0   
763                     6            3        3           2               2   
764                     6            3        3           2               2   
765                     6            3        3           2               2   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1.2  ...  \
0         77.0  28.628792         0            1                   1  ...   
1         52.0  22.506925         0            2                   1  ...   
2         68.0  27.942143         0            2                   1  ...   
3         53.0  24.196494         0            2                   1  ...   
4         52.0  19.100092         0            1                   1  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN         2            3                   1  ...   
762       54.0  19.132653         0            1                   0  ...   
763        NaN        NaN         2            3                   1  ...   
764        NaN        NaN         2            3                   2  ...   
765        NaN        NaN         2            3                   2  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                     0.0         0.0       0.0       0.0                 0   
1                     1.0         0.0       1.0       0.0                 0   
2                     1.0         NaN       NaN       NaN                 0   
3                     0.0         0.0       0.0       0.0                 0   
4                     0.0         0.0       0.0       0.0                 0   
..                    ...         ...       ...       ...               ...   
761                   0.0         0.0       0.0       1.0                 0   
762                   0.0         0.0       1.0       0.0                 0   
763                   0.0         0.0       1.0       0.0                 0   
764                   NaN         NaN       NaN       NaN                 2   
765                   NaN         NaN       NaN       NaN                 2   

     M3_Mental_health_support  m3_illness_events0.1.2  \
0                           0                       2   
1                           0                       2   
2                           0                       3   
3                           0                       3   
4                           0                       2   
..                        ...                     ...   
761                         0                       2   
762                         0                       1   
763                         1                       1   
764                         2                       3   
765                         2                       3   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                            0                            1   
1                            0                            0   
2                            2                            2   
3                            0                            0   
4                            0                            0   
..                         ...                          ...   
761                          0                          

In [68]:
#rename 3 columns which include dots (.) in their names 
df = df.rename(columns={'M0_life_events_1.2': 'M0_life_events_1_2', 'm3_life_events_1.2': 'm3_life_events_1_2', 'm3_illness_events0.1.2':'m3_illness_events0_1_2'})

In [69]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 130 columns):
 #    Column                                                       Dtype  
---   ------                                                       -----  
 0    M0_employment_status                                         int32  
 1    M0_drinking                                                  int32  
 2    M0_diet                                                      int32  
 3    M0_MedDiet                                                   int32  
 4    M0_SpecialDiet                                               int32  
 5    M0_weight                                                    float64
 6    M0_bmi                                                       float64
 7    M0_obese                                                     int32  
 8    M0_exercise                                                  int32  
 9    M0_life_events_1_2                                           in

In [70]:
#convert corresponding int32 feature type to float64
df['M0_employment_status'] = df.M0_employment_status.astype(float)
df['M0_drinking'] = df.M0_drinking.astype(float)
df['M0_diet'] = df.M0_diet.astype(float)
df['M0_MedDiet'] = df.M0_MedDiet.astype(float)
df['M0_SpecialDiet'] = df.M0_SpecialDiet.astype(float)
df['M0_obese'] = df.M0_obese.astype(float)
df['M0_exercise'] = df.M0_exercise.astype(float)
df['M0_life_events_1_2'] = df.M0_life_events_1_2.astype(float)
df['m0_menopausalstatuspre'] = df.m0_menopausalstatuspre.astype(float)
df['M0_psychotropics'] = df.M0_psychotropics.astype(float)
df['m0_Global_QLQ30groups65gl'] = df.m0_Global_QLQ30groups65gl.astype(float)
df['m0_Phys_Fun_QLQ30groups80gl'] = df.m0_Phys_Fun_QLQ30groups80gl.astype(float)
df['m3_life_events_1_2'] = df.m3_life_events_1_2.astype(float)
df['m3_employer_support'] = df.m3_employer_support.astype(float)
df['m3_domestic_help'] = df.m3_domestic_help.astype(float)
df['M3_psychotropics'] = df.M3_psychotropics.astype(float)
df['M3_Mental_health_support'] = df.M3_Mental_health_support.astype(float)
df['m3_illness_events0_1_2'] = df.m3_illness_events0_1_2.astype(float)
df['m3_Global_QLQ30groups65gl'] = df.m3_Global_QLQ30groups65gl.astype(float)
df['m3_Phys_Fun_QLQ30groups80gl'] = df.m3_Phys_Fun_QLQ30groups80gl.astype(float)

In [71]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 130 columns):
 #    Column                                                       Dtype  
---   ------                                                       -----  
 0    M0_employment_status                                         float64
 1    M0_drinking                                                  float64
 2    M0_diet                                                      float64
 3    M0_MedDiet                                                   float64
 4    M0_SpecialDiet                                               float64
 5    M0_weight                                                    float64
 6    M0_bmi                                                       float64
 7    M0_obese                                                     float64
 8    M0_exercise                                                  float64
 9    M0_life_events_1_2                                           fl

In [72]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   6.0          3.0      3.0         2.0             2.0   
762                   4.0          1.0      0.0         0.0             0.0   
763                   6.0          3.0      3.0         2.0             2.0   
764                   6.0          3.0      3.0         2.0             2.0   
765                   6.0          3.0      3.0         2.0             2.0   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0         77.0  28.628792       0.0          1.0                 1.0  ...   
1         52.0  22.506925       0.0          2.0                 1.0  ...   
2         68.0  27.942143       0.0          2.0                 1.0  ...   
3         53.0  24.196494       0.0          2.0                 1.0  ...   
4         52.0  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN       2.0          3.0                 1.0  ...   
762       54.0  19.132653       0.0          1.0                 0.0  ...   
763        NaN        NaN       2.0          3.0                 1.0  ...   
764        NaN        NaN       2.0          3.0                 2.0  ...   
765        NaN        NaN       2.0          3.0                 2.0  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                     0.0         0.0       0.0       0.0               0.0   
1                     1.0         0.0       1.0       0.0               0.0   
2                     1.0         NaN       NaN       NaN               0.0   
3                     0.0         0.0       0.0       0.0               0.0   
4                     0.0         0.0       0.0       0.0               0.0   
..                    ...         ...       ...       ...               ...   
761                   0.0         0.0       0.0       1.0               0.0   
762                   0.0         0.0       1.0       0.0               0.0   
763                   0.0         0.0       1.0       0.0               0.0   
764                   NaN         NaN       NaN       NaN               2.0   
765                   NaN         NaN       NaN       NaN               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                         0.0                     2.0   
1                         0.0                     2.0   
2                         0.0                     3.0   
3                         0.0                     3.0   
4                         0.0                     2.0   
..                        ...                     ...   
761                       0.0                     2.0   
762                       0.0                     1.0   
763                       1.0                     1.0   
764                       2.0                     3.0   
765                       2.0                     3.0   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          0.0                          0.0   
2                          2.0                          2.0   
3                          0.0                          0.0   
4                          0.0                          0.0   
..                         ...                          ...   
761                        0.0                          

In [73]:
#convert corresponding numeric value to NaN
df['M0_employment_status']=df['M0_employment_status'].replace(6, np.NaN)
df['M0_drinking']=df['M0_drinking'].replace(3, np.NaN)
df['M0_diet']=df['M0_diet'].replace(3, np.NaN)
df['M0_MedDiet']=df['M0_MedDiet'].replace(2, np.NaN)
df['M0_SpecialDiet']=df['M0_SpecialDiet'].replace(2, np.NaN)
df['M0_obese']=df['M0_obese'].replace(2, np.NaN)
df['M0_exercise']=df['M0_exercise'].replace(3, np.NaN)
df['M0_life_events_1_2']=df['M0_life_events_1_2'].replace(2, np.NaN)
df['m0_menopausalstatuspre']=df['m0_menopausalstatuspre'].replace(3, np.NaN)
df['M0_psychotropics']=df['M0_psychotropics'].replace(2, np.NaN)
df['m0_Global_QLQ30groups65gl']=df['m0_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m0_Phys_Fun_QLQ30groups80gl']=df['m0_Phys_Fun_QLQ30groups80gl'].replace(2, np.NaN)
df['m3_life_events_1_2']=df['m3_life_events_1_2'].replace(2, np.NaN)
df['m3_employer_support']=df['m3_employer_support'].replace(2, np.NaN)
df['m3_domestic_help']=df['m3_domestic_help'].replace(2, np.NaN)
df['M3_psychotropics']=df['M3_psychotropics'].replace(3, np.NaN)
df['M3_Mental_health_support']=df['M3_Mental_health_support'].replace(2, np.NaN)
df['m3_illness_events0_1_2']=df['m3_illness_events0_1_2'].replace(3, np.NaN)
df['m3_Global_QLQ30groups65gl']=df['m3_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m3_Phys_Fun_QLQ30groups80gl']=df['m3_Phys_Fun_QLQ30groups80gl'].replace(2, np.NaN)

In [74]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   NaN          NaN      NaN         NaN             NaN   
762                   4.0          1.0      0.0         0.0             0.0   
763                   NaN          NaN      NaN         NaN             NaN   
764                   NaN          NaN      NaN         NaN             NaN   
765                   NaN          NaN      NaN         NaN             NaN   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0         77.0  28.628792       0.0          1.0                 1.0  ...   
1         52.0  22.506925       0.0          2.0                 1.0  ...   
2         68.0  27.942143       0.0          2.0                 1.0  ...   
3         53.0  24.196494       0.0          2.0                 1.0  ...   
4         52.0  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN       NaN          NaN                 1.0  ...   
762       54.0  19.132653       0.0          1.0                 0.0  ...   
763        NaN        NaN       NaN          NaN                 1.0  ...   
764        NaN        NaN       NaN          NaN                 NaN  ...   
765        NaN        NaN       NaN          NaN                 NaN  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                     0.0         0.0       0.0       0.0               0.0   
1                     1.0         0.0       1.0       0.0               0.0   
2                     1.0         NaN       NaN       NaN               0.0   
3                     0.0         0.0       0.0       0.0               0.0   
4                     0.0         0.0       0.0       0.0               0.0   
..                    ...         ...       ...       ...               ...   
761                   0.0         0.0       0.0       1.0               0.0   
762                   0.0         0.0       1.0       0.0               0.0   
763                   0.0         0.0       1.0       0.0               0.0   
764                   NaN         NaN       NaN       NaN               2.0   
765                   NaN         NaN       NaN       NaN               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                         0.0                     2.0   
1                         0.0                     2.0   
2                         0.0                     NaN   
3                         0.0                     NaN   
4                         0.0                     2.0   
..                        ...                     ...   
761                       0.0                     2.0   
762                       0.0                     1.0   
763                       1.0                     1.0   
764                       NaN                     NaN   
765                       NaN                     NaN   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          0.0                          0.0   
2                          NaN                          NaN   
3                          0.0                          0.0   
4                          0.0                          0.0   
..                         ...                          ...   
761                        0.0                          

---------

<font color='blue'><b>MICE Data Imoutation</b></font>

In [21]:
imputer = IterativeImputer(imputation_order='ascending',max_iter=10,random_state=42,n_nearest_features=None)

In [22]:
df = imputer.fit_transform(df)

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\sklearn\impute\_iterative.py:699: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


In [23]:
df

array([[ 0.        ,  1.        ,  1.        , ...,  0.        ,
         1.        ,  0.        ],
       [ 0.        ,  1.        ,  1.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 4.        ,  1.        ,  1.        , ..., -0.12738826,
         0.4232174 ,  0.47872838],
       ...,
       [ 0.18249404,  1.19705526,  0.69149857, ...,  0.        ,
         0.        ,  0.08150634],
       [ 1.33734507,  1.18639802,  0.76462203, ...,  0.33454581,
         0.49149813,  0.        ],
       [ 1.32904616,  1.18935458,  0.74933381, ...,  0.33776871,
         0.49437085,  1.        ]])

In [24]:
#Convert Array to Dataframe and name columns
df=pd.DataFrame(df,columns = ['M0_employment_status','M0_drinking','M0_diet','M0_MedDiet','M0_SpecialDiet','M0_weight','M0_bmi',
         'M0_obese','M0_exercise','M0_life_events_1_2','M0_sick_leave_days','M0_extraversion_TIPI',
         'M0_aggreableness_TIPI','M0_conscientiousness_TIPI','M0_emot_stab_TIPI','M0_openess_TIPI',
         'M0_optimism_LOT','M0_comprehensibility_SOC','M0_manageability_SOC','M0_meaningfulness_SOC',
         'M0_Forward_PACT','M0_Trauma_PACT','M0_Flexibility_PACT','M0_self_blame_CERQ','M0_other_blame_CERQ',
         'M0_rumination_CERQ','M0_catastrophizing_CERQ','M0_perspective_CERQ','M0_pos_refus_CERQ',
         'M0_pos_reapp_CERQ','M0_acceptance_CERQ','M0_planning_CERQ','M0_negative_overall_CERQ',
         'M0_positive_overall_CERQ','M0_mindfulness_MAAS','M0_resilience_CDRISC','M0_coping_with_cancer_CBI',
         'M0_fear_of_recur_FCRI','M0_anxiety_HADS','M0_depression_HADS','M0_mental_health_total_HADS',
         'M0_positive_affect_PANAS','M0_negative_affect_PANAS','M0_Global_QLQ30','M0_Phys_Fun_QLQ30',
         'M0_Role_Fun_QLQ30','M0_Emot_Fun_QLQ30','M0_Cogn_Fun_QLQ30','M0_Soc_Fun_QLQ30','M0_Fatigue_QLQ30',
         'M0_Nausea_QLQ30','M0_Pain_QLQ30','M0_Dyspnoea_QLQ30','M0_Insomnia_QLQ30','M0_Apetite_QLQ30',
         'M0_Constipation_QLQ30','M0_Diarrhoea_QLQ30','M0_Financial_QLQ30','M0_Body_Image_BR23',
         'M0_Side_Effects_BR23','M0_Breast_Symptoms_BR23','M0_Arm_Symptoms_BR23','M0_Upset_Hair_Image_BR23',
         'M0_Future_Persp_Image_BR23','M0_Sex_Funct_BR23','M0_Sex_Enjoy_BR23','M0_performancestatus',
         'm0_menopausalstatuspre','M0_surgery','M0_chemo','M0_radio','M0_psychotropics',
         'm0_Global_QLQ30groups65gl','m0_Phys_Fun_QLQ30groups80gl','diagnosis_M0_delay','m3_life_events_1_2',
         'm3_sick_leave_days','m3_employer_support','m3_domestic_help',
         'm3_perceived_suppport_1_item','m3_nccn_distress_thermometer',
         'm3_general_se_1_item','m3_rate_how_much_your_faith_including_religiosity_and_spiri',
         'M3_instumental_support','M3_emotional_support','M3_social_support_total','M3_PTGI_relating_to_others',
         'M3_PTGI_new_possibilities','M3_PTGI_personal_strength','M3_PTGI_spiritual_change','M3_PTGI_appreciation_of_life',
         'M3_PTGI_total_score','M3_negative_affect_PANAS','M3_positive_affect_PANAS','M3_anxiety_HADS',
         'M3_depression_HADS','M3_mental_health_total_HADS','M3_Global_QLQ30','M3_Phys_Fun_QLQ30',
         'M3_Role_Fun_QLQ30','M3_Emot_Fun_QLQ30','M3_Cogn_Fun_QLQ30','M3_Soc_Fun_QLQ30','M3_Fatigue_QLQ30',
         'M3_Nausea_QLQ30','M3_Pain_QLQ30','M3_Dyspnoea_QLQ30','M3_Insomnia_QLQ30','M3_Apetite_QLQ30',
         'M3_Constipation_QLQ30','M3_Diarrhoea_QLQ30','M3_Financial_QLQ30','M3_Body_Image_m3_br23',
         'M3_Side_Effects_m3_br23','M3_Breast_Symptoms_m3_br23','M3_Arm_Symptoms_m3_br23','M3_Upset_Hair_Image_m3_br23',
         'M3_Future_Persp_Image_m3_br23','M3_Sex_Funct_m3_br23','M3_Sex_Enjoy_m3_br23','m3_performancestatus',
         'M3_surgery','M3_chemo','M3_radio','M3_psychotropics','M3_Mental_health_support',
         'm3_illness_events0_1_2','m3_Global_QLQ30groups65gl','m3_Phys_Fun_QLQ30groups80gl','m18_Global_QLQ30groups65gl'])

In [25]:
df

M0_employment_status  M0_drinking   M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                0.000000     1.000000  1.000000    1.000000        0.000000   
1                0.000000     1.000000  1.000000    1.000000        0.000000   
2                4.000000     1.000000  1.000000    1.000000        0.000000   
3                2.000000     2.000000  1.000000    1.000000        0.000000   
4                2.000000     1.000000  0.000000    0.000000        0.000000   
..                    ...          ...       ...         ...             ...   
761              1.636833     1.232720  0.784546    0.573407        0.210972   
762              4.000000     1.000000  0.000000    0.000000        0.000000   
763              0.182494     1.197055  0.691499    0.588745        0.113304   
764              1.337345     1.186398  0.764622    0.633541        0.133956   
765              1.329046     1.189355  0.749334    0.640811        0.109357   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792  0.000000     1.000000            1.000000  ...   
1    52.000000  22.506925  0.000000     2.000000            1.000000  ...   
2    68.000000  27.942143  0.000000     2.000000            1.000000  ...   
3    53.000000  24.196494  0.000000     2.000000            1.000000  ...   
4    52.000000  19.100092  0.000000     1.000000            1.000000  ...   
..         ...        ...       ...          ...                 ...  ...   
761  67.407456  25.576303  0.291472     1.170822            1.000000  ...   
762  54.000000  19.132653  0.000000     1.000000            0.000000  ...   
763  68.651234  23.990233  0.006481     0.879927            1.000000  ...   
764  67.957686  25.492351  0.170502     1.000591            0.675096  ...   
765  68.569990  25.548799  0.173270     1.011751            0.642059  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                0.000000    0.000000  0.000000  0.000000               0.0   
1                1.000000    0.000000  1.000000  0.000000               0.0   
2                1.000000    0.007975  0.295601  0.035748               0.0   
3                0.000000    0.000000  0.000000  0.000000               0.0   
4                0.000000    0.000000  0.000000  0.000000               0.0   
..                    ...         ...       ...       ...               ...   
761              0.000000    0.000000  0.000000  1.000000               0.0   
762              0.000000    0.000000  1.000000  0.000000               0.0   
763              0.000000    0.000000  1.000000  0.000000               0.0   
764              0.356398    0.009122  0.458425  0.038623               2.0   
765              0.398362   -0.001120  0.435050  0.039535               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                    0.000000                2.000000   
1                    0.000000                2.000000   
2                    0.000000                1.664560   
3                    0.000000                1.633951   
4                    0.000000                2.000000   
..                        ...                     ...   
761                  0.000000                2.000000   
762                  0.000000                1.000000   
763                  1.000000                1.000000   
764                  0.228278                1.573243   
765                  0.225122                1.540069   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                     0.000000                     1.000000   
1                     0.000000                     0.000000   
2                    -0.127388                     0.423217   
3                     0.000000                     0.000000   
4                     0.000000                     0.000000   
..                         ...                          ...   
761                   0.000000              

<b>- roundup values</b> (<i>M0_employment_status</i>, <i>M0_drinking</i>, <i>M0_diet</i>, <i>M0_MedDiet</i>,<i>M0_SpecialDiet</i>, <i>M0_obese</i>, <i>M0_exercise</i>, <i>M0_life_events_1_2</i>,<i>m0_menopausalstatuspre</i>, <i>M0_psychotropics</i>, <i>m0_Global_QLQ30groups65gl</i>, <i>m0_Phys_Fun_QLQ30groups80gl</i>,<i>m3_life_events_1_2</i>, <i>m3_employer_support</i>, <i>m3_domestic_help</i>, <i>M3_psychotropics</i>,<i>M3_Mental_health_support</i>, <i>m3_illness_events0_1_2</i>, <i>m3_Global_QLQ30groups65gl</i>, <i>M0_MedDiet</i>,) -

In [26]:
df.M0_employment_status = df.M0_employment_status.round()
df.M0_drinking = df.M0_drinking.round()
df.M0_diet = df.M0_diet.round()
df.M0_MedDiet = df.M0_MedDiet.round()
df.M0_SpecialDiet = df.M0_SpecialDiet.round()
df.M0_obese = df.M0_obese.round()
df.M0_exercise = df.M0_exercise.round()
df.M0_life_events_1_2 = df.M0_life_events_1_2.round()
df.m0_menopausalstatuspre = df.m0_menopausalstatuspre.round()
df.M0_psychotropics = df.M0_psychotropics.round()
df.m0_Global_QLQ30groups65gl = df.m0_Global_QLQ30groups65gl.round()
df.m0_Phys_Fun_QLQ30groups80gl = df.m0_Phys_Fun_QLQ30groups80gl.round()
df.m3_life_events_1_2 = df.m3_life_events_1_2.round()
df.m3_employer_support = df.m3_employer_support.round()
df.m3_domestic_help = df.m3_domestic_help.round()
df.M3_psychotropics = df.M3_psychotropics.round()
df.M3_Mental_health_support = df.M3_Mental_health_support.round()
df.m3_illness_events0_1_2 = df.m3_illness_events0_1_2.round()
df.m3_Global_QLQ30groups65gl = df.m3_Global_QLQ30groups65gl.round()
df.m3_Phys_Fun_QLQ30groups80gl = df.m3_Phys_Fun_QLQ30groups80gl.round()
df.m18_Global_QLQ30groups65gl = df.m18_Global_QLQ30groups65gl.round()

In [27]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   2.0          1.0      1.0         1.0             0.0   
762                   4.0          1.0      0.0         0.0             0.0   
763                   0.0          1.0      1.0         1.0             0.0   
764                   1.0          1.0      1.0         1.0             0.0   
765                   1.0          1.0      1.0         1.0             0.0   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792       0.0          1.0                 1.0  ...   
1    52.000000  22.506925       0.0          2.0                 1.0  ...   
2    68.000000  27.942143       0.0          2.0                 1.0  ...   
3    53.000000  24.196494       0.0          2.0                 1.0  ...   
4    52.000000  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761  67.407456  25.576303       0.0          1.0                 1.0  ...   
762  54.000000  19.132653       0.0          1.0                 0.0  ...   
763  68.651234  23.990233       0.0          1.0                 1.0  ...   
764  67.957686  25.492351       0.0          1.0                 1.0  ...   
765  68.569990  25.548799       0.0          1.0                 1.0  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                0.000000    0.000000  0.000000  0.000000               0.0   
1                1.000000    0.000000  1.000000  0.000000               0.0   
2                1.000000    0.007975  0.295601  0.035748               0.0   
3                0.000000    0.000000  0.000000  0.000000               0.0   
4                0.000000    0.000000  0.000000  0.000000               0.0   
..                    ...         ...       ...       ...               ...   
761              0.000000    0.000000  0.000000  1.000000               0.0   
762              0.000000    0.000000  1.000000  0.000000               0.0   
763              0.000000    0.000000  1.000000  0.000000               0.0   
764              0.356398    0.009122  0.458425  0.038623               2.0   
765              0.398362   -0.001120  0.435050  0.039535               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                         0.0                     2.0   
1                         0.0                     2.0   
2                         0.0                     2.0   
3                         0.0                     2.0   
4                         0.0                     2.0   
..                        ...                     ...   
761                       0.0                     2.0   
762                       0.0                     1.0   
763                       1.0                     1.0   
764                       0.0                     2.0   
765                       0.0                     2.0   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          0.0                          0.0   
2                         -0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   
..                         ...                          ...   
761                        0.0                          

In [28]:
#create dataset based on the 66 most discriminating features
df = df[['diagnosis_M0_delay','m3_sick_leave_days','M3_negative_affect_PANAS','M0_Fatigue_QLQ30','M3_Arm_Symptoms_m3_br23',
        'M3_Arm_Symptoms_m3_br23','m3_employer_support','M0_mindfulness_MAAS','M3_Insomnia_QLQ30','m3_rate_how_much_your_faith_including_religiosity_and_spiri',
        'M0_Arm_Symptoms_BR23','M0_Emot_Fun_QLQ30','M3_Body_Image_m3_br23','M0_negative_affect_PANAS','M3_PTGI_spiritual_change',
        'M3_Role_Fun_QLQ30','M0_chemo','M0_Financial_QLQ30','M0_sick_leave_days','M0_Soc_Fun_QLQ30','m3_performancestatus',
        'M0_life_events_1_2','M0_positive_affect_PANAS','m3_perceived_suppport_1_item','M3_Phys_Fun_QLQ30','M0_rumination_CERQ',
        'M0_Flexibility_PACT','M0_Cogn_Fun_QLQ30','M3_Apetite_QLQ30','M0_Trauma_PACT','M3_instumental_support','M0_negative_overall_CERQ',
        'M0_anxiety_HADS','M0_drinking','M3_chemo','M0_Nausea_QLQ30','M3_mental_health_total_HADS','M3_Future_Persp_Image_m3_br23',
        'm3_nccn_distress_thermometer','m3_life_events_1_2','M0_Apetite_QLQ30','M0_coping_with_cancer_CBI','M0_Phys_Fun_QLQ30',
        'M3_Nausea_QLQ30','M3_Sex_Funct_m3_br23','M0_Forward_PACT','M0_optimism_LOT','M0_Insomnia_QLQ30','M0_Breast_Symptoms_BR23',
        'M0_psychotropics','m0_Global_QLQ30groups65gl','M0_Global_QLQ30','M3_Global_QLQ30','M3_PTGI_appreciation_of_life',
        'M3_Breast_Symptoms_m3_br23','M3_Financial_QLQ30','M3_Upset_Hair_Image_m3_br23','M3_Pain_QLQ30','M0_surgery',
        'M0_pos_refus_CERQ','M3_Soc_Fun_QLQ30','M0_other_blame_CERQ','M0_Sex_Funct_BR23','M3_Mental_health_support',
        'M0_pos_reapp_CERQ','M3_Fatigue_QLQ30','m18_Global_QLQ30groups65gl']]

In [29]:
df

diagnosis_M0_delay  m3_sick_leave_days  M3_negative_affect_PANAS  \
0              3.000000           20.956101                  1.000000   
1              1.000000           47.548774                  1.200000   
2              1.000000            0.000000                  1.502613   
3              2.000000           21.369112                  1.000000   
4              3.000000           39.337434                  2.200000   
..                  ...                 ...                       ...   
761            1.275131           42.750107                  1.400000   
762            0.000000           70.000000                  1.600000   
763            0.000000           90.000000                  3.000000   
764            1.151233           40.444968                  1.808251   
765            1.485362           38.729325                  1.823786   

     M0_Fatigue_QLQ30  M3_Arm_Symptoms_m3_br23  M3_Arm_Symptoms_m3_br23  \
0            0.000000                22.222222                22.222222   
1            0.000000                22.222222                22.222222   
2            0.000000                19.360385                19.360385   
3           22.222222                 0.000000                 0.000000   
4           33.333333                 0.000000                 0.000000   
..                ...                      ...                      ...   
761         24.707715                 0.000000                 0.000000   
762         55.555556                 0.000000                 0.000000   
763         20.836762                11.111111                11.111111   
764         28.480157                13.273926                13.273926   
765         27.926783                13.500930                13.500930   

     m3_employer_support  M0_mindfulness_MAAS  M3_Insomnia_QLQ30  \
0                    1.0             5.133333           0.000000   
1                    1.0             5.133333           0.000000   
2                    0.0             6.000000          23.071746   
3                    1.0             6.000000           0.000000   
4                    1.0             3.800000          33.333333   
..                   ...                  ...                ...   
761                  0.0             4.529424          33.333333   
762                  0.0             4.933333          33.333333   
763                  0.0             5.400000          33.333333   
764                  1.0             4.366991          32.685272   
765                  0.0             4.389331          35.084920   

     m3_rate_how_much_your_faith_including_religiosity_and_spiri  ...  \
0                                            10.000000            ...   
1                                             7.000000            ...   
2                                             7.750623            ...   
3                                            10.000000            ...   
4                                            10.000000            ...   
..                                                 ...            ...   
761                                           7.500000            ...   
762                                           8.790735            ...   
763                                           1.700000            ...   
764                                           5.989063            ...   
765                                           6.092913            ...   

     M3_Pain_QLQ30  M0_surgery  M0_pos_refus_CERQ  M3_Soc_Fun_QLQ30  \
0        33.333333    1.000000           4.500000        100.000000   
1         0.000000    0.000000           2.500000        100.000000   
2        31.436173    0.000000           4.500000         82.368985   
3         0.000000    0.000000           5.000000        100.000000   
4         0.000000    0.000000           4.000000        100.000000   
..             ...         ...                ...               ...   
761      21.457213    0.389837      

----------------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [35]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [36]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df.iloc[:,1:-1].values
Y=df.iloc[:,-1].values

In [37]:
X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]

In [38]:
classifier = SVC(kernel='rbf')

In [39]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train,Y_train)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7763157894736842


<font color='red'>Accuracy: 77.63%</font>

In [40]:
y_predict = classifier.predict(X_test)

In [41]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     51    1
1.0     16    8

For 0: <font color='blue'>98%</font> (51/52) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>33%</font> (8/24) correctly classified -> <font color='red'>poor QoL</font> accuracy

------------

<u><font color='blue'>SMOTE</u></font>

In [50]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [51]:
X_train_smote, Y_train_smote = smote.fit_resample(X_train.astype('float'),Y_train)

In [52]:
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_smote))

Before SMOTE:  Counter({0.0: 469, 1.0: 221})
After SMOTE:  Counter({0.0: 469, 1.0: 469})


In [53]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_smote,Y_train_smote)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7236842105263158


<font color='red'>Accuracy: 72,37%</font>

In [54]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     42   10
1.0     11   13

For 0: <font color='blue'>81%</font> (42/52) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>54%</font> (13/24) correctly classified -> <font color='red'>poor QoL</font> accuracy

-----------------------

<font color='blue'><b><u>without</b></u></font> MICE Data Imoutation <br>
replacing remaining NaN values with the corresponding <u>mean value</u>

In [75]:
df

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   NaN          NaN      NaN         NaN             NaN   
762                   4.0          1.0      0.0         0.0             0.0   
763                   NaN          NaN      NaN         NaN             NaN   
764                   NaN          NaN      NaN         NaN             NaN   
765                   NaN          NaN      NaN         NaN             NaN   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0         77.0  28.628792       0.0          1.0                 1.0  ...   
1         52.0  22.506925       0.0          2.0                 1.0  ...   
2         68.0  27.942143       0.0          2.0                 1.0  ...   
3         53.0  24.196494       0.0          2.0                 1.0  ...   
4         52.0  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761        NaN        NaN       NaN          NaN                 1.0  ...   
762       54.0  19.132653       0.0          1.0                 0.0  ...   
763        NaN        NaN       NaN          NaN                 1.0  ...   
764        NaN        NaN       NaN          NaN                 NaN  ...   
765        NaN        NaN       NaN          NaN                 NaN  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                     0.0         0.0       0.0       0.0               0.0   
1                     1.0         0.0       1.0       0.0               0.0   
2                     1.0         NaN       NaN       NaN               0.0   
3                     0.0         0.0       0.0       0.0               0.0   
4                     0.0         0.0       0.0       0.0               0.0   
..                    ...         ...       ...       ...               ...   
761                   0.0         0.0       0.0       1.0               0.0   
762                   0.0         0.0       1.0       0.0               0.0   
763                   0.0         0.0       1.0       0.0               0.0   
764                   NaN         NaN       NaN       NaN               2.0   
765                   NaN         NaN       NaN       NaN               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                         0.0                     2.0   
1                         0.0                     2.0   
2                         0.0                     NaN   
3                         0.0                     NaN   
4                         0.0                     2.0   
..                        ...                     ...   
761                       0.0                     2.0   
762                       0.0                     1.0   
763                       1.0                     1.0   
764                       NaN                     NaN   
765                       NaN                     NaN   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          0.0                          0.0   
2                          NaN                          NaN   
3                          0.0                          0.0   
4                          0.0                          0.0   
..                         ...                          ...   
761                        0.0                          

In [76]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 130 columns):
 #    Column                                                       Dtype  
---   ------                                                       -----  
 0    M0_employment_status                                         float64
 1    M0_drinking                                                  float64
 2    M0_diet                                                      float64
 3    M0_MedDiet                                                   float64
 4    M0_SpecialDiet                                               float64
 5    M0_weight                                                    float64
 6    M0_bmi                                                       float64
 7    M0_obese                                                     float64
 8    M0_exercise                                                  float64
 9    M0_life_events_1_2                                           fl

<b>- replace NaN values with mean of each column -</b>

In [78]:
df_mean = df.fillna(df.mean())

In [79]:
df_mean

M0_employment_status  M0_drinking   M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                 0.00000     1.000000  1.000000    1.000000        0.000000   
1                 0.00000     1.000000  1.000000    1.000000        0.000000   
2                 4.00000     1.000000  1.000000    1.000000        0.000000   
3                 2.00000     2.000000  1.000000    1.000000        0.000000   
4                 2.00000     1.000000  0.000000    0.000000        0.000000   
..                    ...          ...       ...         ...             ...   
761               1.32672     1.183024  0.781333    0.646667        0.137333   
762               4.00000     1.000000  0.000000    0.000000        0.000000   
763               1.32672     1.183024  0.781333    0.646667        0.137333   
764               1.32672     1.183024  0.781333    0.646667        0.137333   
765               1.32672     1.183024  0.781333    0.646667        0.137333   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792  0.000000     1.000000            1.000000  ...   
1    52.000000  22.506925  0.000000     2.000000            1.000000  ...   
2    68.000000  27.942143  0.000000     2.000000            1.000000  ...   
3    53.000000  24.196494  0.000000     2.000000            1.000000  ...   
4    52.000000  19.100092  0.000000     1.000000            1.000000  ...   
..         ...        ...       ...          ...                 ...  ...   
761  68.737012  25.580845  0.174625     1.034713            1.000000  ...   
762  54.000000  19.132653  0.000000     1.000000            0.000000  ...   
763  68.737012  25.580845  0.174625     1.034713            1.000000  ...   
764  68.737012  25.580845  0.174625     1.034713            0.632928  ...   
765  68.737012  25.580845  0.174625     1.034713            0.632928  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                0.000000     0.00000  0.000000  0.000000               0.0   
1                1.000000     0.00000  1.000000  0.000000               0.0   
2                1.000000     0.00768  0.415152  0.040909               0.0   
3                0.000000     0.00000  0.000000  0.000000               0.0   
4                0.000000     0.00000  0.000000  0.000000               0.0   
..                    ...         ...       ...       ...               ...   
761              0.000000     0.00000  0.000000  1.000000               0.0   
762              0.000000     0.00000  1.000000  0.000000               0.0   
763              0.000000     0.00000  1.000000  0.000000               0.0   
764              0.351931     0.00768  0.415152  0.040909               2.0   
765              0.351931     0.00768  0.415152  0.040909               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                    0.000000                2.000000   
1                    0.000000                2.000000   
2                    0.000000                1.621053   
3                    0.000000                1.621053   
4                    0.000000                2.000000   
..                        ...                     ...   
761                  0.000000                2.000000   
762                  0.000000                1.000000   
763                  1.000000                1.000000   
764                  0.212798                1.621053   
765                  0.212798                1.621053   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                     0.000000                      1.00000   
1                     0.000000                      0.00000   
2                     0.348765                      0.50858   
3                     0.000000                      0.00000   
4                     0.000000                      0.00000   
..                         ...                          ...   
761                   0.000000              

<b>- roundup values</b> (<i>M0_employment_status</i>, <i>M0_drinking</i>, <i>M0_diet</i>, <i>M0_MedDiet</i>,<i>M0_SpecialDiet</i>, <i>M0_obese</i>, <i>M0_exercise</i>, <i>M0_life_events_1_2</i>,<i>m0_menopausalstatuspre</i>, <i>M0_psychotropics</i>, <i>m0_Global_QLQ30groups65gl</i>, <i>m0_Phys_Fun_QLQ30groups80gl</i>,<i>m3_life_events_1_2</i>, <i>m3_employer_support</i>, <i>m3_domestic_help</i>, <i>M3_psychotropics</i>,<i>M3_Mental_health_support</i>, <i>m3_illness_events0_1_2</i>, <i>m3_Global_QLQ30groups65gl</i>, <i>M0_MedDiet</i>,) -

In [80]:
df_mean.M0_employment_status = df_mean.M0_employment_status.round()
df_mean.M0_drinking = df_mean.M0_drinking.round()
df_mean.M0_diet = df_mean.M0_diet.round()
df_mean.M0_MedDiet = df_mean.M0_MedDiet.round()
df_mean.M0_SpecialDiet = df_mean.M0_SpecialDiet.round()
df_mean.M0_obese = df_mean.M0_obese.round()
df_mean.M0_exercise = df_mean.M0_exercise.round()
df_mean.M0_life_events_1_2 = df_mean.M0_life_events_1_2.round()
df_mean.m0_menopausalstatuspre = df_mean.m0_menopausalstatuspre.round()
df_mean.M0_psychotropics = df_mean.M0_psychotropics.round()
df_mean.m0_Global_QLQ30groups65gl = df_mean.m0_Global_QLQ30groups65gl.round()
df_mean.m0_Phys_Fun_QLQ30groups80gl = df_mean.m0_Phys_Fun_QLQ30groups80gl.round()
df_mean.m3_life_events_1_2 = df_mean.m3_life_events_1_2.round()
df_mean.m3_employer_support = df_mean.m3_employer_support.round()
df_mean.m3_domestic_help = df_mean.m3_domestic_help.round()
df_mean.M3_psychotropics = df_mean.M3_psychotropics.round()
df_mean.M3_Mental_health_support = df_mean.M3_Mental_health_support.round()
df_mean.m3_illness_events0_1_2 = df_mean.m3_illness_events0_1_2.round()
df_mean.m3_Global_QLQ30groups65gl = df_mean.m3_Global_QLQ30groups65gl.round()
df_mean.m3_Phys_Fun_QLQ30groups80gl = df_mean.m3_Phys_Fun_QLQ30groups80gl.round()
df_mean.m18_Global_QLQ30groups65gl = df_mean.m18_Global_QLQ30groups65gl.round()

In [81]:
df_mean

M0_employment_status  M0_drinking  M0_diet  M0_MedDiet  M0_SpecialDiet  \
0                     0.0          1.0      1.0         1.0             0.0   
1                     0.0          1.0      1.0         1.0             0.0   
2                     4.0          1.0      1.0         1.0             0.0   
3                     2.0          2.0      1.0         1.0             0.0   
4                     2.0          1.0      0.0         0.0             0.0   
..                    ...          ...      ...         ...             ...   
761                   1.0          1.0      1.0         1.0             0.0   
762                   4.0          1.0      0.0         0.0             0.0   
763                   1.0          1.0      1.0         1.0             0.0   
764                   1.0          1.0      1.0         1.0             0.0   
765                   1.0          1.0      1.0         1.0             0.0   

     M0_weight     M0_bmi  M0_obese  M0_exercise  M0_life_events_1_2  ...  \
0    77.000000  28.628792       0.0          1.0                 1.0  ...   
1    52.000000  22.506925       0.0          2.0                 1.0  ...   
2    68.000000  27.942143       0.0          2.0                 1.0  ...   
3    53.000000  24.196494       0.0          2.0                 1.0  ...   
4    52.000000  19.100092       0.0          1.0                 1.0  ...   
..         ...        ...       ...          ...                 ...  ...   
761  68.737012  25.580845       0.0          1.0                 1.0  ...   
762  54.000000  19.132653       0.0          1.0                 0.0  ...   
763  68.737012  25.580845       0.0          1.0                 1.0  ...   
764  68.737012  25.580845       0.0          1.0                 1.0  ...   
765  68.737012  25.580845       0.0          1.0                 1.0  ...   

     m3_performancestatus  M3_surgery  M3_chemo  M3_radio  M3_psychotropics  \
0                0.000000     0.00000  0.000000  0.000000               0.0   
1                1.000000     0.00000  1.000000  0.000000               0.0   
2                1.000000     0.00768  0.415152  0.040909               0.0   
3                0.000000     0.00000  0.000000  0.000000               0.0   
4                0.000000     0.00000  0.000000  0.000000               0.0   
..                    ...         ...       ...       ...               ...   
761              0.000000     0.00000  0.000000  1.000000               0.0   
762              0.000000     0.00000  1.000000  0.000000               0.0   
763              0.000000     0.00000  1.000000  0.000000               0.0   
764              0.351931     0.00768  0.415152  0.040909               2.0   
765              0.351931     0.00768  0.415152  0.040909               2.0   

     M3_Mental_health_support  m3_illness_events0_1_2  \
0                         0.0                     2.0   
1                         0.0                     2.0   
2                         0.0                     2.0   
3                         0.0                     2.0   
4                         0.0                     2.0   
..                        ...                     ...   
761                       0.0                     2.0   
762                       0.0                     1.0   
763                       1.0                     1.0   
764                       0.0                     2.0   
765                       0.0                     2.0   

     m3_Global_QLQ30groups65gl  m3_Phys_Fun_QLQ30groups80gl  \
0                          0.0                          1.0   
1                          0.0                          0.0   
2                          0.0                          1.0   
3                          0.0                          0.0   
4                          0.0                          0.0   
..                         ...                          ...   
761                        0.0                          

In [82]:
#create dataset based on the 66 most discriminating features
df = df_mean[['diagnosis_M0_delay','m3_sick_leave_days','M3_negative_affect_PANAS','M0_Fatigue_QLQ30','M3_Arm_Symptoms_m3_br23',
        'M3_Arm_Symptoms_m3_br23','m3_employer_support','M0_mindfulness_MAAS','M3_Insomnia_QLQ30','m3_rate_how_much_your_faith_including_religiosity_and_spiri',
        'M0_Arm_Symptoms_BR23','M0_Emot_Fun_QLQ30','M3_Body_Image_m3_br23','M0_negative_affect_PANAS','M3_PTGI_spiritual_change',
        'M3_Role_Fun_QLQ30','M0_chemo','M0_Financial_QLQ30','M0_sick_leave_days','M0_Soc_Fun_QLQ30','m3_performancestatus',
        'M0_life_events_1_2','M0_positive_affect_PANAS','m3_perceived_suppport_1_item','M3_Phys_Fun_QLQ30','M0_rumination_CERQ',
        'M0_Flexibility_PACT','M0_Cogn_Fun_QLQ30','M3_Apetite_QLQ30','M0_Trauma_PACT','M3_instumental_support','M0_negative_overall_CERQ',
        'M0_anxiety_HADS','M0_drinking','M3_chemo','M0_Nausea_QLQ30','M3_mental_health_total_HADS','M3_Future_Persp_Image_m3_br23',
        'm3_nccn_distress_thermometer','m3_life_events_1_2','M0_Apetite_QLQ30','M0_coping_with_cancer_CBI','M0_Phys_Fun_QLQ30',
        'M3_Nausea_QLQ30','M3_Sex_Funct_m3_br23','M0_Forward_PACT','M0_optimism_LOT','M0_Insomnia_QLQ30','M0_Breast_Symptoms_BR23',
        'M0_psychotropics','m0_Global_QLQ30groups65gl','M0_Global_QLQ30','M3_Global_QLQ30','M3_PTGI_appreciation_of_life',
        'M3_Breast_Symptoms_m3_br23','M3_Financial_QLQ30','M3_Upset_Hair_Image_m3_br23','M3_Pain_QLQ30','M0_surgery',
        'M0_pos_refus_CERQ','M3_Soc_Fun_QLQ30','M0_other_blame_CERQ','M0_Sex_Funct_BR23','M3_Mental_health_support',
        'M0_pos_reapp_CERQ','M3_Fatigue_QLQ30','m18_Global_QLQ30groups65gl']]

In [83]:
df

diagnosis_M0_delay  m3_sick_leave_days  M3_negative_affect_PANAS  \
0              3.000000           37.903297                  1.000000   
1              1.000000           37.903297                  1.200000   
2              1.000000            0.000000                  1.784018   
3              2.000000           37.903297                  1.000000   
4              3.000000           37.903297                  2.200000   
..                  ...                 ...                       ...   
761            1.253711           37.903297                  1.400000   
762            0.000000           70.000000                  1.600000   
763            0.000000           90.000000                  3.000000   
764            1.253711           37.903297                  1.784018   
765            1.253711           37.903297                  1.784018   

     M0_Fatigue_QLQ30  M3_Arm_Symptoms_m3_br23  M3_Arm_Symptoms_m3_br23  \
0            0.000000                22.222222                22.222222   
1            0.000000                22.222222                22.222222   
2            0.000000                14.229721                14.229721   
3           22.222222                 0.000000                 0.000000   
4           33.333333                 0.000000                 0.000000   
..                ...                      ...                      ...   
761         28.937067                 0.000000                 0.000000   
762         55.555556                 0.000000                 0.000000   
763         28.937067                11.111111                11.111111   
764         28.937067                14.229721                14.229721   
765         28.937067                14.229721                14.229721   

     m3_employer_support  M0_mindfulness_MAAS  M3_Insomnia_QLQ30  \
0                    0.0             5.133333           0.000000   
1                    1.0             5.133333           0.000000   
2                    0.0             6.000000          36.349693   
3                    0.0             6.000000           0.000000   
4                    0.0             3.800000          33.333333   
..                   ...                  ...                ...   
761                  0.0             4.384749          33.333333   
762                  0.0             4.933333          33.333333   
763                  0.0             5.400000          33.333333   
764                  0.0             4.384749          36.349693   
765                  0.0             4.384749          36.349693   

     m3_rate_how_much_your_faith_including_religiosity_and_spiri  ...  \
0                                            10.000000            ...   
1                                             7.000000            ...   
2                                             6.130521            ...   
3                                            10.000000            ...   
4                                            10.000000            ...   
..                                                 ...            ...   
761                                           7.500000            ...   
762                                           6.130521            ...   
763                                           1.700000            ...   
764                                           6.130521            ...   
765                                           6.130521            ...   

     M3_Pain_QLQ30  M0_surgery  M0_pos_refus_CERQ  M3_Soc_Fun_QLQ30  \
0        33.333333    1.000000           4.500000        100.000000   
1         0.000000    0.000000           2.500000        100.000000   
2        23.286965    0.000000           4.500000         76.190476   
3         0.000000    0.000000           5.000000        100.000000   
4         0.000000    0.000000           4.000000        100.000000   
..             ...         ...                ...               ...   
761      23.286965    0.264228      

---------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [102]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [103]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df_mean.iloc[:,1:-1].values
Y=df_mean.iloc[:,-1].values

In [104]:
X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]

In [105]:
classifier = SVC(kernel='rbf')

In [106]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train,Y_train)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7631578947368421


<font color='red'>Accuracy: 76.32%</font>

In [90]:
y_predict = classifier.predict(X_test)

In [91]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     55    0
1.0     18    3

For 0: <font color='blue'>100%</font> (55/55) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>14%</font> (3/21) correctly classified -> <font color='red'>poor QoL</font> accuracy

-----------

<u><font color='blue'>SMOTE</u></font>

In [92]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [93]:
X_train_smote, Y_train_smote = smote.fit_resample(X_train.astype('float'),Y_train)

In [94]:
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_smote))

Before SMOTE:  Counter({0.0: 517, 1.0: 173})
After SMOTE:  Counter({0.0: 517, 1.0: 517})


In [95]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_smote,Y_train_smote)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7236842105263158


<font color='red'>Accuracy: 72,37%</font>

In [96]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     42   13
1.0      8   13

For 0: <font color='blue'>76%</font> (42/55) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>62%</font> (13/21) correctly classified -> <font color='red'>poor QoL</font> accuracy